### PR TITLE
Make bntest use a file, not bc

### PR DESCRIPTION
**Pull request by richsalz**
Test suite used from boring, written by David Benjamin.
Test driver converted from C++ to C.

### CODE REVIEW

**Comment by richsalz** [comment]
@ekasper , @davidben :  There are a few tests that I commented-out.  Should I hunt them down and fix them?

**Comment by davidben** [comment]
I have to admit, I don't remember what any of these bugfixes are.

It might be worth poking at git log and git blame in the BoringSSL repo to try to find when the tests were added. Most of them probably have corresponding fixes and motivations in the commit message. If there's any you think you want to import, let me know which and I can do the hat-switching approval thing for CLA purposes. I imagine Emilia can do that too.

(Apologies! I try to remember to send fixes your way, but I definitely forget more than I remember. :-( My "would OpenSSL like this or would this just be annoying" heuristic is probably also off.)

Oh, also, our `bn_test` seems smaller than yours. We don't have functions like `BN_GF2m_mod_inv`. Looks like you lost some tests in this PR.

**Comment by richsalz** [comment]
Yes, this is just a WIP to get your test ported.  I'll integrate our existing tests shortly.

**Comment by davidben** [comment]
Ah, yeah. Makes sense.

What's the story with `MAYBE_FIX_TRAILING_GARBAGE_IGNORED`? Does OpenSSL currently not ignore trailing garbage, or are you getting some other result? I don't believe we intentionally changed behavior there, and at a glance OpenSSL has this behavior too, but maybe I'm reading the code wrong or we messed up at some point. That API I think I'd rather match you on behavior.

I think `MAYBE_FIX_NEG0_IS_NEGATIVE` is worth fixing. `MAYBE_FIX_RAND4_FAILURE` appears to be this one:
https://boringssl.googlesource.com/boringssl/+/b3a7b51f18dd192c911c6239e521d7cec62d2340
Looks like I had it silently chop off the bits instead of `BN_R_BITS_TOO_SMALL`, on grounds that `BN_rand` doesn't reject the `bits == 0` when the top or bottom bits are set. *shrug* Don't think it matters much either way.

Oh, by the way, I should have mentioned this earlier, but this format is derived from your `evptests.txt` format. You be able to share a parser with that one. (We split `EVP_PKEY` and `EVP_CIPHER` into separate test binaries, so I needed to pull it into a library when importing.)

(Hilariously, this format has since been incorporated into *ring* with a Rust driver. And I wrote a Go one to verify the BN test vectors. Between your original C parser, our C++ one, and those two, the format's gone a long way.)

**Comment by richsalz** [comment]
The ```GARBAGE``` works one of three times -- for from hex only.  It should always work I think.

I tried that commit for the ```RAND4``` failure and it did not work.

I agree the ``NEG0_IS_NEGATIVE``` should be fixed.  Don't know if it will get done by the release; it's not my area of expertise. :(

I recogonized the file format. :) I'm going to refactor it into a common library later, probably after 1.1

**Comment by richsalz** [comment]
Restored the missing tests.  Now there's just the ```MAYBE_FIX_``` things to deal with :)

**Comment by richsalz** [comment]
Thinking about the ```RAND4``` failure, I prefer what OpenSSL does if only because it matches the doc.  And it's also the least surprising:  you said give me a one-bit number but asked for the top two bits to be set.  :)

**Comment by davidben** [comment]
In that case, the `bits == 0` case should also fail if `top` or `bottom` instruct `bnrand` to set any bits. (I think that was why I went with this behavior. This issue did not seem important enough to find out whether any consumer cared about that case, so I picked the easiest self-consistent option.)

**Comment by richsalz** [comment]
yes, something like this:
```
    if (bits == 0) {
        if (top != -1 || bottom != 0)
            goto toosmall;
        BN_zero(rnd);
        return 1;
    }
    if (bits < 0 || (bits == 1 && top > 0))
        goto toosmall;
```
where:
```
toosmall:
    BNerr(BN_F_BNRAND, BN_R_BITS_TOO_SMALL);
    return NULL;
}
```

**Comment by richsalz** [comment]
updated with proposed fixes for all the issues.

File: crypto/bn/bn_print.c [hunk 1]
```diff
@@ -23,12 +23,9 @@ char *BN_bn2hex(const BIGNUM *a)
...
     char *p;
 
-    if (a->neg && BN_is_zero(a)) {
-        /* "-0" == 3 bytes including NULL terminator */
-        buf = OPENSSL_malloc(3);
-    } else {
-        buf = OPENSSL_malloc(a->top * BN_BYTES * 2 + 2);
-    }
+    if (BN_is_zero(a))
+        return OPENSSL_strdup("0");
```
Comment (levitte) [inline crypto/bn/bn_print.c]: Nit pick: would it make sense to return `"00"`?  Considering we otherwise always give back hex numbers as pairs of nibbles, it seems odd that we'd make an exception for the zero.

Comment (levitte) [inline crypto/bn/bn_print.c]: Thinking again...  never mind that.

**Comment by levitte** [comment]
I'm curious, how was `bntests.txt` generated?  Is there a script that can be used to regenerate it?

**Comment by davidben** [comment]
It was just generated via the old bntest. I just converted the bc lines into this format. The actual sizes and numbers in bntest did not seem of much significance, so I didn't bother documenting it beyond the commit message. I figure if there's ever need for more, they'll just be appended to, not regenerated. A + B = C isn't going to stop being true.
https://boringssl.googlesource.com/boringssl/+log/master/crypto/bn/bn_tests.txt

I did add a tool to check the test vectors against Go's implementation, since that was arguably lost by dropping the `bc` check.
https://boringssl.googlesource.com/boringssl/+/master/crypto/bn/check_bn_tests.go

**Comment by levitte** [comment]
Ok.  Since earlier bntest was proven correct (by comparing with bc or with the perl verifier), that's fine.  Further test vectors should be generated with something *other* than OpenSSL, though, or we won't detect future bugs.  I think this needs to be made very clear, and possibly provide a tool to generate them (in a future PR, I'm not asking for it in this one)

**Comment by davidben** [comment]
Right, hence the verifier. That Go tool verifies the new format and we've been adding to it whenever we added a new test type. This was intended more as a starting point for better tests than an end state. I used the original just to ensure we didn't lose any coverage (in so far as the original provided much coverage. bntests' tests are pretty bad).

I did add a handful of my own, such as these, but there's room for a lot more.
https://boringssl.googlesource.com/boringssl/+/04000e1e312eeb591268946571a684c019228dcd

**Comment by levitte** [comment]
Ah, saw the link there...  I would like that verifier to be part of this PR, and I think `test/` would be the appropriate place.

Not sure how comfortable the lot of us are with Go, so you might see it getting rewritten in Perl some time in the future.  But as long as it's useful in it current form, it should be present.

**Comment by mattcaswell** [comment]
I'd like to push this to the post 1.1.0 milestone. Thoughts?

**Comment by levitte** [comment]
Yes, I think that's a good idea

**Comment by richsalz** [comment]
I think it's good to be in 1.1.0  It's a test, and it's more  comprehensive than what we have now and it did find bugs.

**Comment by ekasper** [comment]
We can always backport tests after the release.

**Comment by richsalz** [comment]
updated doc to be explicit about corner-cases, per @mattcaswell 's request in #1429.

**Comment by davidben** [comment]
By the way, we recently found another negative zero case in BN_div's no_branch codepath. We just fixed it in `bn_correct_top` which seems a reasonable place to do it, but is something we'd rather not diverge on:
https://boringssl.googlesource.com/boringssl/+/899b9b19a4cd3fe526aaf5047ab9234cdca19f7d%5E%21/

I can pull this into a separate PR if you like. I wasn't sure since the other negative zero fixes are in this PR.

**Comment by richsalz** [comment]
Makes sense to include it here.  I'll do it.

**Comment by richsalz** [comment]
Additional commit pushed.  Thanks!

File: crypto/bn/bn_rand.c [hunk 2]
```diff
@@ -20,15 +20,14 @@ static int bnrand(int pseudorand, BIGNUM *rnd, int bits, int top, int bottom)
     int ret = 0, bit, bytes, mask;
     time_t tim;
 
-    if (bits < 0 || (bits == 1 && top > 0)) {
-        BNerr(BN_F_BNRAND, BN_R_BITS_TOO_SMALL);
-        return 0;
-    }
-
     if (bits == 0) {
+        if (top != -1 || bottom != 0)
```
Comment (mattcaswell) [inline crypto/bn/bn_rand.c]: Shouldn't we be using `BN_RAND_TOP_ANY` and `BN_RAND_BOTTOM_ANY` here?

**Comment by mattcaswell** [comment]
Having looked at this in more detail I personally don't feel able to review a change of this size at this stage in the release. I'd be happy to see it go in if someone else steps up to look at it though.

File: crypto/bn/bn_rand.c (hunk 2 above)
Comment (richsalz) [inline crypto/bn/bn_rand.c]: I tried to do that previously, but Viktor -1'd that change because it broke the existing behavior.

Comment (mattcaswell) [inline crypto/bn/bn_rand.c]: How did it break existing behaviour? These are just constants aren't they??

Comment (richsalz) [inline crypto/bn/bn_rand.c]: ah, sorry, yes that's new code.  I was confused and thought you were referring to line 29 and lines 68-79.  I will fix this one if.

Comment (richsalz) [inline crypto/bn/bn_rand.c]: updated commit pushed.

**Comment by richsalz** [comment]
This fixes 6-8 bugs.  Hope someone on the team will reconsider.

**Comment by mattcaswell** [comment]
It's the test that is difficult to review. If you separate them I would review the bug fixes.

**Comment by richsalz** [comment]
But then they are fixes without tests which I don't want to do.

**Comment by mattcaswell** [comment]
@richsalz - this is now the only remaining github issue; git hub pr; rt ticket; gitlab mr or gitlab issue with the 1.1.0 milestone against it. Realistically this isn't going to get merged by tomorrow! Can we push to post-1.1.0 now? :-)

**Comment by richsalz** [comment]
It's your call, @mattcaswell .  Sad Panda.

**Comment by mattcaswell** [comment]
I made my call...pushed to post-1.1.0

**Comment by davidben** [comment]
Keeping things at "Post 1.1.0" the day of the release is certainly still correct, but perhaps it's also worth splitting out the bugfixes with their tests from the rest of the PR? I believe none of those tests depend on `bntests.txt`.

**Comment by levitte** [comment]
Ping?  Time to look at this again?  Perhaps a rebase against a fresher master for starters?

**Comment by richsalz** [comment]
We had discussed this on the team, and drH prefers something based on NCONF.  I'm not going to do that, so this is just sitting here.  I don't want to commit fixes without regression tests, and there's currently nobody working on the NCONF based tests.  So either this goes in as-is and we update the test infrastruture later, or someone does what drH prefers.

If someone says they would approve this, I'll rebase and update this, but as it stands now it's just gonna sit here...

**Comment by levitte** [comment]
You know, I'm looking at this, and am kinda uneasy with the way this is implemented...  not that I think previous methods were much better, but rather looking forward...

The thing that strikes me is that for each new type of expression to test, we need to write an entirely new test function for that specific formula.  While I can understand the immediate gratification of that approach, it feels like a waste of manual cycles in the long run.

So here's a crazy idea following my train of thought; what would you say about a test driven approach with a generated bntest.c.  Just as a mental exercise, say we have a bntests.txt format looking like this:

```
Expression = c590e57ee64fced3ca84d4bb013bba7d633e68b2ff4e27bf1db43f386dbfcce501f112b7fff6fb9436a576ccfccce12867becf02b91961453ea41f414764407d + 0
Result = c590e57ee64fced3ca84d4bb013bba7d633e68b2ff4e27bf1db43f386dbfcce501f112b7fff6fb9436a576ccfccce12867becf02b91961453ea41f414764407d
```

And have all expressions we want to test to be of that type, no matter the complexity of the formula.  modmul, modexp and other similar calculations can be recognised fairly easily, or one might create "functions" for them to be part of an expression.

Either way, from such a set of tests, it shouldn't be too hard to generate the series of BN calls to implement them and get a result, and for any expression lacking a corresponding result, it should be quite easy to feed the expression (possibly with a bit of massage) to scripts for `bc`, `perl`, `go` or what have you and get the result generated (possibly through more than one, just to check that we get the same result everywhere).

**Comment by richsalz** [comment]
I think you'll get the same feedback: use NCONF.  I'm not going to be able to do that.
This scheme uses the same fileformat as evptest.

**Comment by richsalz** [comment]
And also, I prefer a datafile-based approach, rather than building a custom program when there is really no need for that.

**Comment by levitte** [comment]
Not sure if you meant the same thing as me when you're talking datafile-based approach, @richsalz.  Do you?

**Comment by levitte** [comment]
I'm not sure what a switch to NCONF would give us...

**Comment by richsalz** [comment]
I am not sure what NCONF buys us either, except it's on less custom file parser, but it's what drH wants and I deferred to him.  But I do want the tests and bugfixes in this PR, so if you're interested in it, I'll update it.  As for file-based, I mean that there is no reason to "create" bntest.c, it should just read a data file with test data, like evptest, ssltest, etc.

**Comment by levitte** [comment]
> As for file-based, I mean that there is no reason to "create" bntest.c, it should just read a data file with test data, like evptest, ssltest, etc.

Sure.  But then, I'd rather have it parse expressions and calculate them than have one specialised test function for every kind of expression we want to check.  Kinda what `bc.pl` is doing...

**Comment by richsalz** [comment]
Enh, don't care.  But the current approach easily maps testcase to "call this API"

**Comment by davidben** [comment]
The problem with using expressions is that there are multiple functions for a given mathematical operation with different performance and timing properties. You want to test all of them. For instance, test `TestModMul` tests `BN_mod_mul` and, if odd, `BN_mod_mul_montgomery`. Some operations are also more complex than standard mathematical operations. With the same example, if I write "a * b (mod c)", do I want to test `BN_mul` with `BN_nmod` or `BN_mod_mul`?

You are, after all, a crypto library, not a calculator.

**Comment by levitte** [comment]
Mm, yeah, I hear you...  I guess that my main concern is keeping things in sync, and it would be nice to automate result generation using bc or perl, so there will essentially be two places where each test type will have to be codified.

I guess my past experience with keeping three different build systems in sync has me get hiccups when thinking about another case of keeping things in sync...

**Comment by davidben** [comment]
That's true. Having an easy to parse common format helps. The Go checker we have duplicates the parser, but that won't have to change much. Adding a new test type to it is very little code.

**Comment by ekasper** [comment]
I really don't see what NCONF would buy us, it's overly cumbersome, especially for this simple test. I also don't see value in a complicated expression parser - as @davidben says we're not a calculator, we need tests for the simple primitives (BN_add, BN_mul etc) and it's fine to have a small custom test routine per each of those. I think the file test approach is perfectly fine for the basic unit test.

I do, however, see value in running randomized tests and cross-checking against a different implementation. Bugs have been found this way. However this should be set up as a separate, heavier stress test (effectively a type of fuzzer). The current unit test doesn't really stress the code enough to find corner-case bugs, so I'm not too sad to lose this.

So to sum up, I'd like to see this PR done. It definitely makes adding regression tests for specific inputs easier. But do consider adding an independent checker script that generates/verifies the results to ensure that when we check in new test vectors, we first verify them with something other than OpenSSL.

**Comment by richsalz** [comment]
Rebased and push -f'd.
I'll look at the idea of writing a test to call, say bc.  I hope doing that won't be in vain.

**Comment by levitte** [comment]
> But do consider adding an independent checker script that generates/verifies the results to ensure that when we check in new test vectors, we first verify them with something other than OpenSSL.

`bc.pl` can easily be retrofitted for this. If noone else will, I can.

**Comment by levitte** [comment]
... and now I saw @richsalz's comment...  have at it!

**Comment by ekasper** [comment]
Could you adjust this to use testutil.h, and also rename the methods to use our conventions (`test_sum` instead of `TestSum`) etc?

**Comment by ekasper** [comment]
Also, why do you keep

```
/*
 * Things in boring, not in openssl
 */
#define HAVE_HEXDUMP 0
#define HAVE_BN_PADDED 0
#define HAVE_BN_SQRT 0
```

Do you have plans to add support for these? Otherwise this is dead code.

Apologies, Github won't actually let me leave line comments because the diff is too large. I'd ask you to split it up to convert one test at a time but it's probably too onerous for you - so perhaps just give the PR another lick to remove debugging leftovers and make it comply with our style and conventions.

**Comment by richsalz** [comment]
Yes I want to add the missing HAVE_xxx features at some point.
I will convert to use the test_util.h
I will rename the tests
I will convert bc.pl to feed into bc to *run* the tests as a verification (@levitte, if you want to help with that, I would appreciate it)
Anything else?

**Comment by richsalz** [comment]
@ekasper do you want one test for the whole file (as it is now), or one test for each test_xxx routine?

**Comment by levitte** [comment]
Err, regarding the idea of having `bc.pl` feed things to `bc`, I don't quite understand. `bc.pl` is a simple `bc` replacement...

**Comment by richsalz** [comment]
Of course you're right, @levitte.  I have a parser written, working on the rest.

**Comment by ekasper** [comment]
One per test_xxx would be more informative if it's not too much trouble.

File: test/bntests.pl [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+my $infile = shift || 'bntests.txt';
+die "No such file, $infile" unless -f $infile;
+open my $IN, $infile || die "Can't read $infile, $!\n";
+
+my %stanza = ();
+my $l = 0;
+while ( <$IN> ) {
+    $l++;
+    chop;
```
Comment (levitte) [inline test/bntests.pl]: This isn't the best function to use.  General recommend used to be to use `chomp` instead, but even that doesn't always hold, and our past experience shows that this is the best choice:

``` perl
    s|\R$||;
```

File: test/bntests.pl [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    next if /^#/;
```
Comment (levitte) [inline test/bntests.pl]: Better regexp is `/^\s*#/`

File: test/bntests.pl [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    if ( /^$/ ) {
```
Comment (levitte) [inline test/bntests.pl]: Better regexp: `/^\s*$/`

File: test/bntests.pl [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        if ( keys %stanza ) {
+            &evaluate($l, %stanza);
+            %stanza = ();
+        }
+        next;
+    }
```
Comment (levitte) [inline test/bntests.pl]: May I suggest trimming start and ending spaces here?

``` perl
s|^\s+||;
s|\s+$||;
```

File: test/bntests.pl [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        my $quotient = &bn($s{'Quotient'});
+        my $remainder = &bn($s{'Remainder'});
+        my $a = &bn($s{'A'});
+        my $b = &bn($s{'B'});
+
+        # First the remainder test.
+        $b->bmul($quotient);
+        my $rempassed = $remainder == $a->bsub($b) ? 1 : 0;
+
+        # XXX TODO BROKEN
```
Comment (levitte) [inline test/bntests.pl]: Which do you mean is broken in this regard, what Math::BigInt or OpenSSL BN?  Mathematically, Math::BigInt is more correct, but OpenSSL BN behaves the same way as bc, and may be what people expect after all these years...  So the question is what you'd like to see changed.  I'm thinking you should mention that explicitely.

File: test/bntests.pl [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    my ($sign, $hex) = ($x =~ /^([+\-]?)(.*)$/);
+
+    $hex = '0x' . $hex if $hex !~ /^0x/;
+    return Math::BigInt->from_hex($sign.$hex);
+}
+
+sub evaluate()
+{
+    my ( $lineno ) = shift;
+    my ( %s ) = @_;
```
Comment (levitte) [inline test/bntests.pl]: For both lines, parentheses aren't really needed.

File: test/bntests.pl [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    if ( defined $s{'Sum'} ) {
+        # Sum = A + B
+        my $sum = &bn($s{'Sum'});
```
Comment (levitte) [inline test/bntests.pl]: Function calls do not need to be preceded with `&` any more.  That's perl 4... or was it even perl 3?

**Comment by levitte** [comment]
I like the `bntests.pl` script.  Clean and very readable, just a few style and regexp comments...

One possible wishlist item was if it could take stanza like this:

```
Sum = ?
A = 568af72ec870
B = aa44
```

and whenever it does, simply give us its answer instead of trying to verify.  That would make this an excellent tool to generate answers when we come up with new tests.  Make life easier and all that...

**Comment by richsalz** [comment]
The "Sum=?" is an interesting idea, but I hope it's not required for this PR. :)  I also wonder if someone won't just type equations into python or dc or whatever...
Running this on my machine takes as long as almost the entire rest of the test suite.  It's not something you'd run every time.   I need to figure out why one set of tests is failing.

**Comment by levitte** [comment]
No, not a requirement.

**Comment by levitte** [comment]
You need to rebase on a fresher master.  The build issues you see in Travis were fixed just recently.

**Comment by levitte** [comment]
> I need to figure out why one set of tests is failing.

You mean that `bntests.pl` fails the ModSqrt tests, right?  I'm noticing that with ModSqrt, the modulo hack doesn't seem to be needed.  It seems that (specifically) the BN_mod_sqrt() does the modulo part mathematically correctly....  It's BN_mod() that has a more unusual behavior.  So that hack that you have in there might not be needed.  I'm currently checking to confirm.

**Comment by levitte** [comment]
Errrr, I can't make heads or tails of this...  I'm starting to doubt some of the stanzas, for example:

```
ModSqrt = 3
A = -5
P = 7
```

In my head, -5 (mod 7) = 2, this stanza seems to imply that -5 (mod 7) = 9, and that makes no sense to me.

**Comment by levitte** [comment]
Never mind my previous comment...  just had to re-wrap my brain around what ModSqrt does...

**Comment by levitte** [comment]
I just had a much closer look at `TestModSqrt()`, and I've questions about [these lines](https://github.com/openssl/openssl/blob/7da7a426afa8a2a7004490fee8f81ba4f33d4479/test/bntest.c#L1427-L1433).  From the look of it with the `BN_sub` operation, it seems to check (P - ModSqrt)^2 = A (mod P), not the ModSqrt = A (mod P) as the comment in `bntests.txt` says.  @davidben, would you mind commenting on that?

**Comment by levitte** [comment]
As for the ModSqrt check in `bntests.pl`, it doesn't quite do the job.

1.  The modulo hack isn't needed, because BN_mod_sqrt() uses BN_nnmod(), which *is* mathematically correct.
2.  The ModSqrt operation isn't just a straight sqrt(A % P), you also have to find the nearest equivalent integer square root up (uhmmm, not sure I expressed that correctly, but [this page](http://math.stackexchange.com/questions/633160/modular-arithmetic-find-the-square-root) explains the math quite thoroughly).  Coding that in perl might not be quite as easy as it looks...

**Comment by levitte** [comment]
There may be a reason why the old `bntest.c` never actually output any bc operation for ModSqrt...  all things considered, I'm thinking `bntests.pl` should simply skip over those, for the same kind of reason.

**Comment by davidben** [comment]
> I just had a much closer look at TestModSqrt(), and I've questions about these lines. From the look of it with the BN_sub operation, it seems to check (P - ModSqrt)^2 = A (mod P), not the ModSqrt = A (mod P) as the comment in bntests.txt says. @davidben, would you mind commenting on that?

I assume "ModSqrt = A (mod P)" was a typo for "ModSqrt^2 = A (mod P)"? So, if you just simplify what you wrote mod P:

(P - ModSqrt)^2 = (-ModSqrt)^2 = ModSqrt^2 (mod P)

`BN_mod_sqrt` computes a modular square root of its input (supposing the input is a quadratic residue and the modulus is a prime). Like with integers, there are two square roots, X and -X (also known as P - X). Returning either from `BN_mod_sqrt` is legal. (Hrm, now that I look at it, it seems that code incorrectly allows `BN_mod_sqrt` to return P instead of zero. Like with integers, zero only has one square root because 0 = -0.)

Although, if you all are trying to write a checker, the original code in BoringSSL already came with one:
https://boringssl.googlesource.com/boringssl/+/master/crypto/bn/check_bn_tests.go#230
As done there, you can check it by squaring instead of taking the modular square root. Although then you can't do your "ModSqrt = ?" thing.

(But this checker is in Go which even provides [ModSqrt](https://golang.org/pkg/math/big/#Int.ModSqrt). I forget why I decided to do it this way.)

File: test/bntests.pl (hunk 9 above)
Comment (richsalz) [inline test/bntests.pl]: when i tried that:
```
; perl --version

This is perl 5, version 18, subversion 2 (v5.18.2) built for x86_64-linux-gnu-thread-multi
(with 44 registered patches, see perl -V for more detail)

Copyright 1987-2013, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

; perl -cw test/bntests.pl
Too many arguments for main::bn at test/bntests.pl line 30, near "})"
Too many arguments for main::bn at test/bntests.pl line 31, near "})"
Too many arguments for main::bn at test/bntests.pl line 32, near "})"
test/bntests.pl had compilation errors.
        exit 255
```

File: test/bntests.pl (hunk 7 above)
Comment (richsalz) [inline test/bntests.pl]: outdated comment; removed the TODO/broken part.

File: test/bntests.pl (hunk 4 above)
Comment (richsalz) [inline test/bntests.pl]: no, that's not the fileformat that bntest.c expects.

File: test/bntests.pl (hunk 6 above)
Comment (richsalz) [inline test/bntests.pl]: that's not what bntest.c expects.

File: test/bntests.pl [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+# Run the tests specified in bntests.txt, as a check against OpenSSL.
+use strict;
+use warnings;
+use Math::BigInt;
+
+my $EXPECTED_FAILURES = 53;
+my $failures = 0;
+
+sub bn()
```
Comment (levitte) [inline test/bntests.pl]: With the `()`, you have declared that `bn` takes zero parameters, but when you use it, you violate that contract.

You can change that by either removing the parens:

``` perl
sub bn
```

or by declaring it takes one scalar argument:

``` perl
sub bn($)
```

This is the reason you were "forced" to use `&` in all calls.

**Comment by levitte** [comment]
> > I just had a much closer look at TestModSqrt(), and I've questions about these lines. From the look of it with the BN_sub operation, it seems to check (P - ModSqrt)^2 = A (mod P), not the ModSqrt = A (mod P) as the comment in bntests.txt says. @davidben, would you mind commenting on that?
>
> I assume "ModSqrt = A (mod P)" was a typo for "ModSqrt^2 = A (mod P)"?

Yes.  Typo...

> So, if you just simplify what you wrote mod P:
>
> (P - ModSqrt)^2 = (-ModSqrt)^2 = ModSqrt^2 (mod P)
>
> `BN_mod_sqrt` computes a modular square root of its input (supposing the input is a quadratic residue and the modulus is a prime). Like with integers, there are two square roots, X and -X (also known as P - X). Returning either from `BN_mod_sqrt` is legal. (Hrm, now that I look at it, it seems that code incorrectly allows `BN_mod_sqrt` to return P instead of zero. Like with integers, zero only has one square root because 0 = -0.)

True, and now I see that you do compare against both.

> Although, if you all are trying to write a checker, the original code in BoringSSL already came with one:
https://boringssl.googlesource.com/boringssl/+/master/crypto/bn/check_bn_tests.go#230
As done there, you can check it by squaring instead of taking the modular square root. Although then you can't do your "ModSqrt = ?" thing.

Cool...

**Comment by richsalz** [comment]
@levitte , @ekasper I believe I have addressed all the open issues, ready for review.

File: test/bntests.pl [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        my $modsqrt = &bn($s{'ModSqrt'});
+        my $a = &bn($s{'A'});
+        my $p = &bn($s{'P'});
+
+        # With OpenSSL's BN, the result of -10 % 3 is -1
+        # while Math::BigInt, the result is 2.
+        # Math::BigInt does something different with a negative modulus,
+        # while OpenSSL's BN and bc treat it like a positive number...
+        $a->babs();
+        $p->babs();
```
Comment (levitte) [inline test/bntests.pl]: If you remove the `babs` calls, the expected failures are down at 50.  Like I discussed already, `BN_mod_sqrt()` uses `BN_nnmod()`, which *does* behave like `Math::BigInt` modulo, so no extra hacking needed here.

**Comment by levitte** [comment]
@richsalz:
> The modsqrt comparison in bntests.pl *does do* a M*M=(A mod P). 

Yes, but that's not always that easy. Take the example that I mentioned earlier, with ModSqrt =3, A = -5 and P =7. Would you agree that A mod P = 2 in this case? If so, please explain to me how ModSqrt^2 can equal 2.

The interesting thing is that `BN_mod_sqrt()` returns 4 (or, I assume, -4) in this case.

**Comment by davidben** [comment]
> Yes, but that's not always that easy. Take the example that I mentioned earlier, with ModSqrt =3, A = -5 and P =7. Would you agree that A mod P = 2 in this case? If so, please explain to me how ModSqrt^2 can equal 2.

3 * 3 = 9 = 2 (mod 7)

> The interesting thing is that BN_mod_sqrt() returns 4 (or, I assume, -4) in this case.

4 * 4 = 16 = 2 (mod 7). Also -4 = 3 (mod 7). 3 and 4 are the square roots of 2 mod 7.

**Comment by levitte** [comment]
> 3 * 3 = 9 = 2 (mod 7)

Yes.  @richsalz's script isn't doing it mod7, though...

**Comment by richsalz** [comment]
I fixed it to do ModSqrt mod P and now there are only six failures.

**Comment by levitte** [comment]
This patch takes it down to zero failures:

``` diff
diff --git a/test/bntests.pl b/test/bntests.pl
index d927f21..79bb01d 100755
--- a/test/bntests.pl
+++ b/test/bntests.pl
@@ -11,7 +11,7 @@ use strict;
 use warnings;
 use Math::BigInt;
 
-my $EXPECTED_FAILURES = 6;
+my $EXPECTED_FAILURES = 0;
 my $failures = 0;
 
 sub bn
@@ -112,13 +112,6 @@ sub evaluate
         $modsqrt->bmul($modsqrt);
         my $a = bn($s{'A'});
         my $p = bn($s{'P'});
-
-        # With OpenSSL's BN, the result of -10 % 3 is -1
-        # while Math::BigInt, the result is 2.
-        # Math::BigInt does something different with a negative modulus,
-        # while OpenSSL's BN and bc treat it like a positive number...
-        $a->babs();
-        $p->babs();
         $a->bmod($p);
         $modsqrt->bmod($p);
         return if $modsqrt == $a;
```

**Comment by richsalz** [comment]
... and I think we're finally done.  All open issues resolved.

**Comment by levitte** [comment]
Not quite. Did you see @ekasper's comments on code style?

**Comment by richsalz** [comment]
yes, she wanted the functions renamed and use the test harness.  I did that.

**Comment by levitte** [comment]
There's still a bit of CamelCase in `bnstest.c`...  Our coding style is a liiiittle vague regarding that, but it's not exactly something we usually do.

**Comment by richsalz** [comment]
okay, changed 3 function.  But "getBN" should be okay, since it's an openssl datatype.
also, leading/trailing whitespace around key and value *are* stripped, so i edited the perl tester.

**Comment by levitte** [comment]
We have `parsebn` and `parsedecbn` just above those...  let's have a consistent style.

**Comment by richsalz** [comment]
okAY, i fixED thoSE tWO functioNS.

**Comment by levitte** [comment]
Other way than I expected, but ok ;-)

**Comment by richsalz** [comment]
commit 56f05b9 on master.  Thanks for the extensive reviews, @levitte @ekasper; it definitely improved the tests!
